<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.54">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Tutorial 1: GenomicRanges and range-based analyses – Bioc2024: Interoperability between R and Python using BiocPy</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href=".././assets/short.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Tutorial 1: <code>GenomicRanges</code> and range-based analyses</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src=".././assets/logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      <div class="sidebar-tools-main">
    <a href="https://github.com/BiocPy/" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
</div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Interoperability between R and Python using BiocPy</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Tutorials</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../tutorials/genomic_ranges.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tutorial 1: <code>GenomicRanges</code> and range-based analyses</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../tutorials/annotate_cell_types.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tutorial 2: Access single-cell datasets from <code>scRNAseq</code> collection and annotate cell types</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../tutorials/sessioninfo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Session Info</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#outline" id="toc-outline" class="nav-link active" data-scroll-target="#outline">Outline</a></li>
  <li><a href="#prerequisites" id="toc-prerequisites" class="nav-link" data-scroll-target="#prerequisites">Prerequisites</a>
  <ul class="collapse">
  <li><a href="#installation" id="toc-installation" class="nav-link" data-scroll-target="#installation">Installation</a></li>
  </ul></li>
  <li><a href="#save-annotations-as-rds" id="toc-save-annotations-as-rds" class="nav-link" data-scroll-target="#save-annotations-as-rds">1. Save Annotations as RDS</a></li>
  <li><a href="#reading-rds-files-in-python" id="toc-reading-rds-files-in-python" class="nav-link" data-scroll-target="#reading-rds-files-in-python">2. Reading RDS files in Python</a></li>
  <li><a href="#basic-genomic-operations" id="toc-basic-genomic-operations" class="nav-link" data-scroll-target="#basic-genomic-operations">2. Basic Genomic Operations</a>
  <ul class="collapse">
  <li><a href="#create-a-genomicrangeslist-by-gene" id="toc-create-a-genomicrangeslist-by-gene" class="nav-link" data-scroll-target="#create-a-genomicrangeslist-by-gene">2.1 Create a <code>GenomicRangesList</code> by gene</a></li>
  <li><a href="#finding-transcription-start-sites-tss" id="toc-finding-transcription-start-sites-tss" class="nav-link" data-scroll-target="#finding-transcription-start-sites-tss">2.2 Finding Transcription Start Sites (TSS)</a></li>
  <li><a href="#defining-promoter-regions" id="toc-defining-promoter-regions" class="nav-link" data-scroll-target="#defining-promoter-regions">2.3 Defining Promoter Regions</a></li>
  </ul></li>
  <li><a href="#overlap-analysis" id="toc-overlap-analysis" class="nav-link" data-scroll-target="#overlap-analysis">3. Overlap Analysis</a>
  <ul class="collapse">
  <li><a href="#reading-chip-seq-peaks" id="toc-reading-chip-seq-peaks" class="nav-link" data-scroll-target="#reading-chip-seq-peaks">3.1 Reading ChIP-seq Peaks</a></li>
  <li><a href="#finding-overlaps-with-tss" id="toc-finding-overlaps-with-tss" class="nav-link" data-scroll-target="#finding-overlaps-with-tss">3.2 Finding Overlaps with TSS</a></li>
  <li><a href="#finding-overlaps-with-promoters" id="toc-finding-overlaps-with-promoters" class="nav-link" data-scroll-target="#finding-overlaps-with-promoters">3.3 Finding Overlaps with Promoters</a></li>
  <li><a href="#finding-overlaps-with-exons" id="toc-finding-overlaps-with-exons" class="nav-link" data-scroll-target="#finding-overlaps-with-exons">3.4 Finding Overlaps with Exons</a></li>
  </ul></li>
  <li><a href="#advanced-operations" id="toc-advanced-operations" class="nav-link" data-scroll-target="#advanced-operations">4. Advanced Operations</a>
  <ul class="collapse">
  <li><a href="#comparing-exonic-vs.-intronic-binding" id="toc-comparing-exonic-vs.-intronic-binding" class="nav-link" data-scroll-target="#comparing-exonic-vs.-intronic-binding">4.1 Comparing Exonic vs.&nbsp;Intronic Binding</a></li>
  <li><a href="#finding-overlaps-with-the-first-exon" id="toc-finding-overlaps-with-the-first-exon" class="nav-link" data-scroll-target="#finding-overlaps-with-the-first-exon">4.2 Finding Overlaps with the first exon</a></li>
  <li><a href="#resizing-and-shifting-peaks" id="toc-resizing-and-shifting-peaks" class="nav-link" data-scroll-target="#resizing-and-shifting-peaks">4.3 Resizing and Shifting Peaks</a></li>
  </ul></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises">5. Exercises</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/BiocPy/BiocWorkshop2024/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Tutorial 1: <code>GenomicRanges</code> and range-based analyses</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Genomic range operations are fundamental to many bioinformatics analyses. They allow us to work with intervals of genomic coordinates, which is crucial for understanding the relationships between different genomic features such as genes, regulatory elements, and experimental data like ChIP-seq peaks. In this tutorial, we’ll explore how to work with genomic interval data using BiocPy’s <a href="https://github.com/BiocPy/GenomicRanges/">GenomicRanges</a> package, which provides Python implementations similar to the R/Bioconductor <a href="https://github.com/Bioconductor/GenomicRanges">GenomicRanges package</a>.</p>
<section id="outline" class="level2">
<h2 class="anchored" data-anchor-id="outline">Outline</h2>
<p>In this workshop, we’ll walk through several key aspects of working with genomic ranges in Python:</p>
<ol type="1">
<li>Reading Genomic Data: We’ll start by reading in genomic data from RDS files, including exon positions grouped by transcripts.</li>
<li>Basic Genomic Operations: We’ll cover fundamental operations like finding transcription start sites (TSS) and promoter regions.</li>
<li>Overlap Analysis: We’ll learn how to find overlaps between different genomic features, a common task in many analyses.</li>
<li>Advanced Operations: We’ll explore more complex operations like finding peaks within specific regions and resizing genomic intervals.</li>
</ol>
</section>
<section id="prerequisites" class="level2">
<h2 class="anchored" data-anchor-id="prerequisites">Prerequisites</h2>
<p>Before we begin, please ensure that you have the following packages installed:</p>
<section id="installation" class="level3">
<h3 class="anchored" data-anchor-id="installation">Installation</h3>
<p>Let’s start by installing the required packages for R and Python.</p>
<section id="python-shell" class="level4">
<h4 class="anchored" data-anchor-id="python-shell">Python (shell)</h4>
<p>You can install the Python packages using pip:</p>
<div id="cell-4" class="cell" data-scrolled="true" data-execution_count="22">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip install <span class="op">-</span>U biocutils genomicranges rds2py numpy pandas geniml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Requirement already satisfied: biocutils in /Users/kancherj/miniforge3/lib/python3.10/site-packages (0.1.5)
Requirement already satisfied: genomicranges in /Users/kancherj/miniforge3/lib/python3.10/site-packages (0.4.28)
Requirement already satisfied: rds2py in /Users/kancherj/miniforge3/lib/python3.10/site-packages (0.4.4)
Requirement already satisfied: numpy in /Users/kancherj/miniforge3/lib/python3.10/site-packages (1.26.4)
Collecting numpy
  Using cached numpy-2.0.0-cp310-cp310-macosx_14_0_arm64.whl.metadata (60 kB)
Requirement already satisfied: pandas in /Users/kancherj/miniforge3/lib/python3.10/site-packages (2.2.2)
Requirement already satisfied: geniml in /Users/kancherj/miniforge3/lib/python3.10/site-packages (0.4.0)
Requirement already satisfied: biocframe&gt;=0.5.11 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from genomicranges) (0.5.11)
Requirement already satisfied: iranges&gt;=0.2.11 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from iranges[optional]&gt;=0.2.11-&gt;genomicranges) (0.2.11)
Requirement already satisfied: Cython in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from rds2py) (3.0.10)
Requirement already satisfied: scipy in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from rds2py) (1.12.0)
Requirement already satisfied: singlecellexperiment&gt;=0.4.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from rds2py) (0.4.6)
Requirement already satisfied: summarizedexperiment&gt;=0.4.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from rds2py) (0.4.5)
Requirement already satisfied: python-dateutil&gt;=2.8.2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from pandas) (2.9.0.post0)
Requirement already satisfied: pytz&gt;=2020.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from pandas) (2024.1)
Requirement already satisfied: tzdata&gt;=2022.7 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from pandas) (2024.1)
Requirement already satisfied: anndata&gt;0.9.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.10.8)
Requirement already satisfied: genimtools&gt;=0.0.12 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.0.13)
Requirement already satisfied: fastembed&gt;=0.2.5 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.3.4)
Requirement already satisfied: gensim in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (4.3.2)
Requirement already satisfied: huggingface-hub in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.23.5)
Requirement already satisfied: logmuse in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.2.8)
Requirement already satisfied: matplotlib in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (3.8.3)
Requirement already satisfied: hnswlib in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.8.0)
Requirement already satisfied: paramiko&gt;=3.0.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (3.4.0)
Requirement already satisfied: peppy&gt;=0.40.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.40.3)
Requirement already satisfied: pyBigWig in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.3.23)
Requirement already satisfied: qdrant-client in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (1.10.1)
Requirement already satisfied: requests&gt;=2.31.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (2.31.0)
Requirement already satisfied: scanpy in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (1.10.2)
Requirement already satisfied: seaborn in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.13.2)
Requirement already satisfied: torch in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (2.3.1)
Requirement already satisfied: ubiquerg&gt;=0.6.3 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.8.0)
Requirement already satisfied: pyarrow in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (15.0.2)
Requirement already satisfied: lightning in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (2.3.3)
Requirement already satisfied: langchain-huggingface==0.0.2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.0.2)
Requirement already satisfied: botocore&gt;=1.34.54 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (1.34.131)
Requirement already satisfied: boto3&gt;=1.34.54 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (1.34.131)
Requirement already satisfied: pybiocfilecache&gt;=0.4.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (0.4.0)
Requirement already satisfied: zarr&gt;=2.17.2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (2.18.2)
Requirement already satisfied: pyyaml&gt;=6.0.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (6.0.1)
Requirement already satisfied: s3fs&gt;=2024.3.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from geniml) (2024.6.1)
Requirement already satisfied: langchain-core&lt;0.3,&gt;=0.1.52 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langchain-huggingface==0.0.2-&gt;geniml) (0.2.20)
Requirement already satisfied: sentence-transformers&gt;=2.6.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langchain-huggingface==0.0.2-&gt;geniml) (3.0.1)
Requirement already satisfied: text-generation&lt;0.8.0,&gt;=0.7.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langchain-huggingface==0.0.2-&gt;geniml) (0.7.0)
Requirement already satisfied: tokenizers&gt;=0.19.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langchain-huggingface==0.0.2-&gt;geniml) (0.19.1)
Requirement already satisfied: transformers&gt;=4.39.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langchain-huggingface==0.0.2-&gt;geniml) (4.42.4)
Requirement already satisfied: array-api-compat!=1.5,&gt;1.4 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from anndata&gt;0.9.0-&gt;geniml) (1.7.1)
Requirement already satisfied: exceptiongroup in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from anndata&gt;0.9.0-&gt;geniml) (1.2.1)
Requirement already satisfied: h5py&gt;=3.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from anndata&gt;0.9.0-&gt;geniml) (3.11.0)
Requirement already satisfied: natsort in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from anndata&gt;0.9.0-&gt;geniml) (8.4.0)
Requirement already satisfied: packaging&gt;=20.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from anndata&gt;0.9.0-&gt;geniml) (23.2)
Requirement already satisfied: jmespath&lt;2.0.0,&gt;=0.7.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from boto3&gt;=1.34.54-&gt;geniml) (1.0.1)
Requirement already satisfied: s3transfer&lt;0.11.0,&gt;=0.10.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from boto3&gt;=1.34.54-&gt;geniml) (0.10.2)
Requirement already satisfied: urllib3!=2.2.0,&lt;3,&gt;=1.25.4 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from botocore&gt;=1.34.54-&gt;geniml) (2.2.1)
Requirement already satisfied: PyStemmer&lt;3.0.0,&gt;=2.2.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from fastembed&gt;=0.2.5-&gt;geniml) (2.2.0.1)
Requirement already satisfied: loguru&lt;0.8.0,&gt;=0.7.2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from fastembed&gt;=0.2.5-&gt;geniml) (0.7.2)
Requirement already satisfied: mmh3&lt;5.0,&gt;=4.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from fastembed&gt;=0.2.5-&gt;geniml) (4.1.0)
Requirement already satisfied: onnx&lt;2.0.0,&gt;=1.15.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from fastembed&gt;=0.2.5-&gt;geniml) (1.16.1)
Requirement already satisfied: onnxruntime&lt;2.0.0,&gt;=1.17.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from fastembed&gt;=0.2.5-&gt;geniml) (1.18.1)
Requirement already satisfied: pillow&lt;11.0.0,&gt;=10.3.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from fastembed&gt;=0.2.5-&gt;geniml) (10.4.0)
Requirement already satisfied: snowballstemmer&lt;3.0.0,&gt;=2.2.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from fastembed&gt;=0.2.5-&gt;geniml) (2.2.0)
Requirement already satisfied: tqdm&lt;5.0,&gt;=4.66 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from fastembed&gt;=0.2.5-&gt;geniml) (4.66.2)
Requirement already satisfied: filelock in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from huggingface-hub-&gt;geniml) (3.15.4)
Requirement already satisfied: fsspec&gt;=2023.5.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from huggingface-hub-&gt;geniml) (2024.6.1)
Requirement already satisfied: typing-extensions&gt;=3.7.4.3 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from huggingface-hub-&gt;geniml) (4.10.0)
Requirement already satisfied: ncls==0.0.68 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from iranges&gt;=0.2.11-&gt;iranges[optional]&gt;=0.2.11-&gt;genomicranges) (0.0.68)
Requirement already satisfied: polars in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from iranges[optional]&gt;=0.2.11-&gt;genomicranges) (1.1.0)
Requirement already satisfied: bcrypt&gt;=3.2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from paramiko&gt;=3.0.0-&gt;geniml) (4.1.3)
Requirement already satisfied: cryptography&gt;=3.3 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from paramiko&gt;=3.0.0-&gt;geniml) (42.0.8)
Requirement already satisfied: pynacl&gt;=1.5 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from paramiko&gt;=3.0.0-&gt;geniml) (1.5.0)
Requirement already satisfied: rich&gt;=10.3.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from peppy&gt;=0.40.1-&gt;geniml) (13.7.1)
Requirement already satisfied: pephubclient&gt;=0.4.2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from peppy&gt;=0.40.1-&gt;geniml) (0.4.2)
Requirement already satisfied: sqlalchemy&lt;2.1,&gt;=2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from pybiocfilecache&gt;=0.4.0-&gt;geniml) (2.0.31)
Requirement already satisfied: six&gt;=1.5 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from python-dateutil&gt;=2.8.2-&gt;pandas) (1.16.0)
Requirement already satisfied: charset-normalizer&lt;4,&gt;=2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from requests&gt;=2.31.0-&gt;geniml) (3.3.2)
Requirement already satisfied: idna&lt;4,&gt;=2.5 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from requests&gt;=2.31.0-&gt;geniml) (3.6)
Requirement already satisfied: certifi&gt;=2017.4.17 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from requests&gt;=2.31.0-&gt;geniml) (2024.2.2)
Requirement already satisfied: aiobotocore&lt;3.0.0,&gt;=2.5.4 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from s3fs&gt;=2024.3.1-&gt;geniml) (2.13.1)
Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from s3fs&gt;=2024.3.1-&gt;geniml) (3.9.5)
Requirement already satisfied: asciitree in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from zarr&gt;=2.17.2-&gt;geniml) (0.3.3)
Requirement already satisfied: numcodecs&gt;=0.10.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from zarr&gt;=2.17.2-&gt;geniml) (0.13.0)
Requirement already satisfied: fasteners in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from zarr&gt;=2.17.2-&gt;geniml) (0.19)
Requirement already satisfied: smart-open&gt;=1.8.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from gensim-&gt;geniml) (7.0.4)
Requirement already satisfied: lightning-utilities&lt;2.0,&gt;=0.10.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from lightning-&gt;geniml) (0.11.5)
Requirement already satisfied: torchmetrics&lt;3.0,&gt;=0.7.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from lightning-&gt;geniml) (1.4.0.post0)
Requirement already satisfied: pytorch-lightning in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from lightning-&gt;geniml) (2.3.3)
Requirement already satisfied: sympy in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from torch-&gt;geniml) (1.13.0)
Requirement already satisfied: networkx in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from torch-&gt;geniml) (3.3)
Requirement already satisfied: jinja2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from torch-&gt;geniml) (3.1.3)
Requirement already satisfied: setuptools in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from logmuse-&gt;geniml) (69.2.0)
Requirement already satisfied: contourpy&gt;=1.0.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from matplotlib-&gt;geniml) (1.2.0)
Requirement already satisfied: cycler&gt;=0.10 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from matplotlib-&gt;geniml) (0.12.1)
Requirement already satisfied: fonttools&gt;=4.22.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from matplotlib-&gt;geniml) (4.50.0)
Requirement already satisfied: kiwisolver&gt;=1.3.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from matplotlib-&gt;geniml) (1.4.5)
Requirement already satisfied: pyparsing&gt;=2.3.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from matplotlib-&gt;geniml) (3.1.2)
Requirement already satisfied: grpcio&gt;=1.41.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from qdrant-client-&gt;geniml) (1.64.1)
Requirement already satisfied: grpcio-tools&gt;=1.41.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from qdrant-client-&gt;geniml) (1.64.1)
Requirement already satisfied: httpx&gt;=0.20.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from httpx[http2]&gt;=0.20.0-&gt;qdrant-client-&gt;geniml) (0.27.0)
Requirement already satisfied: portalocker&lt;3.0.0,&gt;=2.7.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from qdrant-client-&gt;geniml) (2.10.1)
Requirement already satisfied: pydantic&gt;=1.10.8 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from qdrant-client-&gt;geniml) (2.8.2)
Requirement already satisfied: joblib in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (1.4.2)
Requirement already satisfied: legacy-api-wrap&gt;=1.4 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (1.4)
Requirement already satisfied: numba&gt;=0.56 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (0.60.0)
Requirement already satisfied: patsy in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (0.5.6)
Requirement already satisfied: pynndescent&gt;=0.5 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (0.5.13)
Requirement already satisfied: scikit-learn&gt;=0.24 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (1.5.1)
Requirement already satisfied: session-info in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (1.0.0)
Requirement already satisfied: statsmodels&gt;=0.13 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (0.14.2)
Requirement already satisfied: umap-learn!=0.5.0,&gt;=0.5 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scanpy-&gt;geniml) (0.5.6)
Requirement already satisfied: wrapt&lt;2.0.0,&gt;=1.10.10 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from aiobotocore&lt;3.0.0,&gt;=2.5.4-&gt;s3fs&gt;=2024.3.1-&gt;geniml) (1.16.0)
Requirement already satisfied: aioitertools&lt;1.0.0,&gt;=0.5.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from aiobotocore&lt;3.0.0,&gt;=2.5.4-&gt;s3fs&gt;=2024.3.1-&gt;geniml) (0.11.0)
Requirement already satisfied: aiosignal&gt;=1.1.2 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1-&gt;s3fs&gt;=2024.3.1-&gt;geniml) (1.3.1)
Requirement already satisfied: attrs&gt;=17.3.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1-&gt;s3fs&gt;=2024.3.1-&gt;geniml) (23.2.0)
Requirement already satisfied: frozenlist&gt;=1.1.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1-&gt;s3fs&gt;=2024.3.1-&gt;geniml) (1.4.1)
Requirement already satisfied: multidict&lt;7.0,&gt;=4.5 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1-&gt;s3fs&gt;=2024.3.1-&gt;geniml) (6.0.5)
Requirement already satisfied: yarl&lt;2.0,&gt;=1.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1-&gt;s3fs&gt;=2024.3.1-&gt;geniml) (1.9.4)
Requirement already satisfied: async-timeout&lt;5.0,&gt;=4.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1-&gt;s3fs&gt;=2024.3.1-&gt;geniml) (4.0.3)
Requirement already satisfied: cffi&gt;=1.12 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from cryptography&gt;=3.3-&gt;paramiko&gt;=3.0.0-&gt;geniml) (1.16.0)
Requirement already satisfied: protobuf&lt;6.0dev,&gt;=5.26.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from grpcio-tools&gt;=1.41.0-&gt;qdrant-client-&gt;geniml) (5.27.2)
Requirement already satisfied: anyio in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from httpx&gt;=0.20.0-&gt;httpx[http2]&gt;=0.20.0-&gt;qdrant-client-&gt;geniml) (4.4.0)
Requirement already satisfied: httpcore==1.* in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from httpx&gt;=0.20.0-&gt;httpx[http2]&gt;=0.20.0-&gt;qdrant-client-&gt;geniml) (1.0.5)
Requirement already satisfied: sniffio in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from httpx&gt;=0.20.0-&gt;httpx[http2]&gt;=0.20.0-&gt;qdrant-client-&gt;geniml) (1.3.1)
Requirement already satisfied: h11&lt;0.15,&gt;=0.13 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from httpcore==1.*-&gt;httpx&gt;=0.20.0-&gt;httpx[http2]&gt;=0.20.0-&gt;qdrant-client-&gt;geniml) (0.14.0)
Requirement already satisfied: h2&lt;5,&gt;=3 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from httpx[http2]&gt;=0.20.0-&gt;qdrant-client-&gt;geniml) (4.1.0)
Requirement already satisfied: jsonpatch&lt;2.0,&gt;=1.33 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langchain-core&lt;0.3,&gt;=0.1.52-&gt;langchain-huggingface==0.0.2-&gt;geniml) (1.33)
Requirement already satisfied: langsmith&lt;0.2.0,&gt;=0.1.75 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langchain-core&lt;0.3,&gt;=0.1.52-&gt;langchain-huggingface==0.0.2-&gt;geniml) (0.1.88)
Requirement already satisfied: tenacity!=8.4.0,&lt;9.0.0,&gt;=8.1.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langchain-core&lt;0.3,&gt;=0.1.52-&gt;langchain-huggingface==0.0.2-&gt;geniml) (8.2.3)
Requirement already satisfied: llvmlite&lt;0.44,&gt;=0.43.0dev0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from numba&gt;=0.56-&gt;scanpy-&gt;geniml) (0.43.0)
Requirement already satisfied: coloredlogs in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from onnxruntime&lt;2.0.0,&gt;=1.17.0-&gt;fastembed&gt;=0.2.5-&gt;geniml) (15.0.1)
Requirement already satisfied: flatbuffers in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from onnxruntime&lt;2.0.0,&gt;=1.17.0-&gt;fastembed&gt;=0.2.5-&gt;geniml) (24.3.25)
Requirement already satisfied: typer&gt;=0.7.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from pephubclient&gt;=0.4.2-&gt;peppy&gt;=0.40.1-&gt;geniml) (0.12.3)
Requirement already satisfied: annotated-types&gt;=0.4.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from pydantic&gt;=1.10.8-&gt;qdrant-client-&gt;geniml) (0.7.0)
Requirement already satisfied: pydantic-core==2.20.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from pydantic&gt;=1.10.8-&gt;qdrant-client-&gt;geniml) (2.20.1)
Requirement already satisfied: markdown-it-py&gt;=2.2.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from rich&gt;=10.3.0-&gt;peppy&gt;=0.40.1-&gt;geniml) (3.0.0)
Requirement already satisfied: pygments&lt;3.0.0,&gt;=2.13.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from rich&gt;=10.3.0-&gt;peppy&gt;=0.40.1-&gt;geniml) (2.17.2)
Requirement already satisfied: threadpoolctl&gt;=3.1.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from scikit-learn&gt;=0.24-&gt;scanpy-&gt;geniml) (3.5.0)
Requirement already satisfied: regex!=2019.12.17 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from transformers&gt;=4.39.0-&gt;langchain-huggingface==0.0.2-&gt;geniml) (2024.5.15)
Requirement already satisfied: safetensors&gt;=0.4.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from transformers&gt;=4.39.0-&gt;langchain-huggingface==0.0.2-&gt;geniml) (0.4.3)
Requirement already satisfied: MarkupSafe&gt;=2.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from jinja2-&gt;torch-&gt;geniml) (2.1.5)
Requirement already satisfied: stdlib-list in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from session-info-&gt;scanpy-&gt;geniml) (0.10.0)
Requirement already satisfied: mpmath&lt;1.4,&gt;=1.1.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from sympy-&gt;torch-&gt;geniml) (1.3.0)
Requirement already satisfied: pycparser in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from cffi&gt;=1.12-&gt;cryptography&gt;=3.3-&gt;paramiko&gt;=3.0.0-&gt;geniml) (2.21)
Requirement already satisfied: humanfriendly&gt;=9.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from coloredlogs-&gt;onnxruntime&lt;2.0.0,&gt;=1.17.0-&gt;fastembed&gt;=0.2.5-&gt;geniml) (10.0)
Requirement already satisfied: hyperframe&lt;7,&gt;=6.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from h2&lt;5,&gt;=3-&gt;httpx[http2]&gt;=0.20.0-&gt;qdrant-client-&gt;geniml) (6.0.1)
Requirement already satisfied: hpack&lt;5,&gt;=4.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from h2&lt;5,&gt;=3-&gt;httpx[http2]&gt;=0.20.0-&gt;qdrant-client-&gt;geniml) (4.0.0)
Requirement already satisfied: jsonpointer&gt;=1.9 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from jsonpatch&lt;2.0,&gt;=1.33-&gt;langchain-core&lt;0.3,&gt;=0.1.52-&gt;langchain-huggingface==0.0.2-&gt;geniml) (2.4)
Requirement already satisfied: orjson&lt;4.0.0,&gt;=3.9.14 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from langsmith&lt;0.2.0,&gt;=0.1.75-&gt;langchain-core&lt;0.3,&gt;=0.1.52-&gt;langchain-huggingface==0.0.2-&gt;geniml) (3.10.6)
Requirement already satisfied: mdurl~=0.1 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from markdown-it-py&gt;=2.2.0-&gt;rich&gt;=10.3.0-&gt;peppy&gt;=0.40.1-&gt;geniml) (0.1.2)
Requirement already satisfied: click&gt;=8.0.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from typer&gt;=0.7.0-&gt;pephubclient&gt;=0.4.2-&gt;peppy&gt;=0.40.1-&gt;geniml) (8.1.7)
Requirement already satisfied: shellingham&gt;=1.3.0 in /Users/kancherj/miniforge3/lib/python3.10/site-packages (from typer&gt;=0.7.0-&gt;pephubclient&gt;=0.4.2-&gt;peppy&gt;=0.40.1-&gt;geniml) (1.5.4)</code></pre>
</div>
</div>
</section>
<section id="r" class="level4">
<h4 class="anchored" data-anchor-id="r">R</h4>
<pre><code>```r
BiocManager::install(c("AnnotationHub"), 
     repos='http://cran.us.r-project.org')
```</code></pre>
</section>
</section>
</section>
<section id="save-annotations-as-rds" class="level2">
<h2 class="anchored" data-anchor-id="save-annotations-as-rds">1. Save Annotations as RDS</h2>
<p>Let’s download the human reference genome and save the exon positions grouped by transcripts. For the purpose of the tutorial, we’ll limit the exons to chromosome 22.</p>
<section id="r-1" class="level4">
<h4 class="anchored" data-anchor-id="r-1">R</h4>
<pre><code>```r
suppressMessages(library(AnnotationHub))

ah &lt;- AnnotationHub()
ensdb &lt;- query(ah, "Ensembl 112 EnsDb for Homo sapiens")[[1]]
exons_by_tx &lt;- exonsBy(ensdb, 
    by = "tx", filter = SeqNameFilter(c("22")), 
    columns= c("exon_id", "tx_name", "tx_id", "gene_name", "gene_id"))
saveRDS(exons_by_tx, "hg38_exons_by_tx.rds")
```</code></pre>
</section>
</section>
<section id="reading-rds-files-in-python" class="level2">
<h2 class="anchored" data-anchor-id="reading-rds-files-in-python">2. Reading RDS files in Python</h2>
<p>The <a href="https://github.com/biocpy/rds2py">rds2py</a> Python package allows us to read RDS files and create equivalent Python representations of R objects. Key features include:</p>
<ul>
<li>Parsing common R objects into Python equivalents (e.g., matrices to NumPy arrays, data frames to Pandas DataFrames).</li>
<li>Ability to read S4 classes, enabling direct parsing of Bioconductor data types from R to Python.</li>
</ul>
<p>Reading an RDS file with rds2py involves two steps:</p>
<ol type="1">
<li>Parse the RDS file into a Python dictionary containing data, its attributes, and associated metadata.</li>
<li>Convert this dictionary into a suitable Python object using specific parser functions.</li>
</ol>
<p>This process allows a seamless transition between R and Python for bioinformatics analyses.</p>
<div id="cell-8" class="cell" data-scrolled="true" data-execution_count="23">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> rds2py <span class="im">import</span> read_rds</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>hg38_robject <span class="op">=</span> read_rds(<span class="st">"./hg38_exons_by_tx.rds"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> rds2py.granges <span class="im">import</span> as_granges_list</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>by_tx <span class="op">=</span> as_granges_list(hg38_robject)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Exons by transcript:"</span>)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(by_tx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Exons by transcript:
GenomicRangesList with 5387 ranges and 0 metadata columns
 
Name: ENST00000006251 
GenomicRanges with 9 ranges and 6 metadata columns
    seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
[0]    chr22 44677057 - 44677241               + | ENSE00001838743 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         1
[1]    chr22 44702492 - 44702609               + | ENSE00003647870 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         2
[2]    chr22 44714591 - 44714672               + | ENSE00003614159 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         3
[3]    chr22 44725244 - 44725293               + | ENSE00003568825 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         4
[4]    chr22 44726577 - 44726635               + | ENSE00003465556 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         5
[5]    chr22 44731730 - 44731822               + | ENSE00003642381 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         6
[6]    chr22 44732251 - 44732392               + | ENSE00003658491 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         7
[7]    chr22 44735027 - 44735163               + | ENSE00003692865 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         8
[8]    chr22 44736772 - 44737681               + | ENSE00001846334 ENST00000006251 ENST00000006251      PRR5 ENSG00000186654         9
------
seqinfo(1 sequences): chr22
 
Name: ENST00000008876 
GenomicRanges with 10 ranges and 6 metadata columns
    seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
[0]    chr22 50603133 - 50603499               + | ENSE00003608148 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         1
[1]    chr22 50603626 - 50603720               + | ENSE00003768317 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         2
[2]    chr22 50603841 - 50605065               + | ENSE00003772801 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         3
[3]    chr22 50605368 - 50605444               + | ENSE00003773674 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         4
[4]    chr22 50605562 - 50605735               + | ENSE00003765641 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         5
[5]    chr22 50605825 - 50605935               + | ENSE00003763622 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         6
[6]    chr22 50606658 - 50606766               + | ENSE00003773228 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         7
[7]    chr22 50606921 - 50606992               + | ENSE00003769486 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         8
[8]    chr22 50610212 - 50610311               + | ENSE00003772161 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735         9
[9]    chr22 50610707 - 50613982               + | ENSE00003731955 ENST00000008876 ENST00000008876  MAPK8IP2 ENSG00000008735        10
------
seqinfo(1 sequences): chr22
 
Name: ENST00000043402 
GenomicRanges with 2 ranges and 6 metadata columns
    seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
[0]    chr22 20268071 - 20268319               - | ENSE00001358408 ENST00000043402 ENST00000043402     RTN4R ENSG00000040608         1
[1]    chr22 20241415 - 20243111               - | ENSE00001557601 ENST00000043402 ENST00000043402     RTN4R ENSG00000040608         2
------
seqinfo(1 sequences): chr22
 
Name:  
GenomicRanges with 15 ranges and 6 metadata columns
     seqnames              ranges          strand        exon_id   tx_name     tx_id gene_name gene_id exon_rank
        &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;         &lt;list&gt;    &lt;list&gt;    &lt;list&gt;    &lt;list&gt;  &lt;list&gt;    &lt;list&gt;
 [0]    chr22 33919995 - 33920477               - |  LRG_856t1e1 LRG_856t2 LRG_856t2    LARGE1 LRG_856         1
 [1]    chr22 33761371 - 33761559               - |  LRG_856t1e3 LRG_856t2 LRG_856t2    LARGE1 LRG_856         2
 [2]    chr22 33650367 - 33650669               - |  LRG_856t1e4 LRG_856t2 LRG_856t2    LARGE1 LRG_856         3
          ...                 ...             ... |          ...       ...       ...       ...     ...       ...
[12]    chr22 33283202 - 33283349               - | LRG_856t1e14 LRG_856t2 LRG_856t2    LARGE1 LRG_856        13
[13]    chr22 33277060 - 33277256               - | LRG_856t1e15 LRG_856t2 LRG_856t2    LARGE1 LRG_856        14
[14]    chr22 33272509 - 33274625               - | LRG_856t1e16 LRG_856t2 LRG_856t2    LARGE1 LRG_856        15
------
seqinfo(1 sequences): chr22
 
Name: LRG_856t2 
GenomicRanges with 7 ranges and 6 metadata columns
    seqnames              ranges          strand      exon_id  tx_name    tx_id gene_name gene_id exon_rank
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;       &lt;list&gt;   &lt;list&gt;   &lt;list&gt;    &lt;list&gt;  &lt;list&gt;    &lt;list&gt;
[0]    chr22 37244114 - 37244266               - | LRG_97t1e1 LRG_97t1 LRG_97t1      RAC2  LRG_97         1
[1]    chr22 37241587 - 37241659               - | LRG_97t1e2 LRG_97t1 LRG_97t1      RAC2  LRG_97         2
[2]    chr22 37232801 - 37232919               - | LRG_97t1e3 LRG_97t1 LRG_97t1      RAC2  LRG_97         3
[3]    chr22 37231932 - 37231995               - | LRG_97t1e4 LRG_97t1 LRG_97t1      RAC2  LRG_97         4
[4]    chr22 37231231 - 37231391               - | LRG_97t1e5 LRG_97t1 LRG_97t1      RAC2  LRG_97         5
[5]    chr22 37226671 - 37226804               - | LRG_97t1e6 LRG_97t1 LRG_97t1      RAC2  LRG_97         6
[6]    chr22 37225270 - 37226040               - | LRG_97t1e7 LRG_97t1 LRG_97t1      RAC2  LRG_97         7
------
seqinfo(1 sequences): chr22
 
Name: LRG_97t1 
GenomicRanges with 21 ranges and 6 metadata columns
     seqnames              ranges          strand        exon_id   tx_name     tx_id gene_name gene_id exon_rank
        &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;         &lt;list&gt;    &lt;list&gt;    &lt;list&gt;    &lt;list&gt;  &lt;list&gt;    &lt;list&gt;
 [0]    chr22 20982297 - 20982572               + |  LRG_989t1e1 LRG_989t1 LRG_989t1     LZTR1 LRG_989         1
 [1]    chr22 20983027 - 20983090               + |  LRG_989t1e2 LRG_989t1 LRG_989t1     LZTR1 LRG_989         2
 [2]    chr22 20985841 - 20985898               + |  LRG_989t1e3 LRG_989t1 LRG_989t1     LZTR1 LRG_989         3
          ...                 ...             ... |          ...       ...       ...       ...     ...       ...
[18]    chr22 20996696 - 20996802               + | LRG_989t1e19 LRG_989t1 LRG_989t1     LZTR1 LRG_989        19
[19]    chr22 20996886 - 20996967               + | LRG_989t1e20 LRG_989t1 LRG_989t1     LZTR1 LRG_989        20
[20]    chr22 20997232 - 20999033               + | LRG_989t1e21 LRG_989t1 LRG_989t1     LZTR1 LRG_989        21
------
seqinfo(1 sequences): chr22
 
</code></pre>
</div>
</div>
</section>
<section id="basic-genomic-operations" class="level2">
<h2 class="anchored" data-anchor-id="basic-genomic-operations">2. Basic Genomic Operations</h2>
<p>Now, let’s perform some basic operations like finding Transcription Start Sites (TSS) and promoter regions. These operations are fundamental in genomic analysis as they help us identify key regulatory regions of the genome.</p>
<section id="create-a-genomicrangeslist-by-gene" class="level3">
<h3 class="anchored" data-anchor-id="create-a-genomicrangeslist-by-gene">2.1 Create a <code>GenomicRangesList</code> by gene</h3>
<p>To identify TSS or define promoter regions, let’s first reprocess the input to create a <code>GenomicRangesList</code> by gene symbols.</p>
<p>To achieve this, we unlist the <code>GenomicRangesList</code> object. This is accomplished in Python using the <code>as_genomic_ranges()</code> method.</p>
<div id="cell-10" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>all_ranges <span class="op">=</span> by_tx.as_genomic_ranges()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we split the object using the <code>gene_name</code> metadata column in <code>mcols()</code>.</p>
<pre class="{important}"><code>We provide accessors to get or set attributes of the class. Most folks in Python may be familiar with direct access to class members (via properties or @property), but this should generally be avoided, as it is too easy to perform modifications via one-liners with the class.property on the left-hand side of an assignment.

For more information, please refer to our [developer guide](https://github.com/BiocPy/developer_guide).</code></pre>
<pre class="{note}"><code>While gene IDs are unique, gene symbols are not. In addition, this list has genes with no symbols.</code></pre>
<div id="cell-12" class="cell" data-scrolled="true" data-execution_count="25">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>by_gene <span class="op">=</span> all_ranges.split(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    groups<span class="op">=</span>all_ranges.get_mcols().get_column(<span class="st">"gene_name"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Exons by gene:"</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(by_gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Exons by gene:
GenomicRangesList with 932 ranges and 0 metadata columns
 
Name:  
GenomicRanges with 1846 ranges and 6 metadata columns
                seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
                   &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
ENST00000255890    chr22 23402361 - 23402568               - | ENSE00001754779 ENST00000255890 ENST00000255890           ENSG00000290920         1
ENST00000255890    chr22 23402037 - 23402156               - | ENSE00001700441 ENST00000255890 ENST00000255890           ENSG00000290920         2
ENST00000255890    chr22 23401841 - 23401954               - | ENSE00004028898 ENST00000255890 ENST00000255890           ENSG00000290920         3
                     ...                 ...             ... |             ...             ...             ...       ...             ...       ...
ENST00000706202    chr22 21008247 - 21009450               - | ENSE00003995059 ENST00000706202 ENST00000706202           ENSG00000291240         7
ENST00000714325    chr22 18939446 - 18947693               + | ENSE00003802171 ENST00000714325 ENST00000714325           ENSG00000284294         1
ENST00000715281    chr22 35838583 - 35838658               - | ENSE00004026430 ENST00000715281 ENST00000715281           ENSG00000293594         1
------
seqinfo(1 sequences): chr22
 
Name: 5_8S_rRNA 
GenomicRanges with 1 range and 6 metadata columns
                seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
                   &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
ENST00000612732    chr22 11249809 - 11249960               - | ENSE00003735240 ENST00000612732 ENST00000612732 5_8S_rRNA ENSG00000276871         1
------
seqinfo(1 sequences): chr22
 
Name: A4GALT 
GenomicRanges with 20 ranges and 6 metadata columns
                seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
                   &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
ENST00000249005    chr22 42695491 - 42695634               - | ENSE00001109573 ENST00000249005 ENST00000249005    A4GALT ENSG00000128274         1
ENST00000249005    chr22 42692122 - 42693998               - | ENSE00000880583 ENST00000249005 ENST00000249005    A4GALT ENSG00000128274         2
ENST00000381278    chr22 42720797 - 42720820               - | ENSE00003818792 ENST00000381278 ENST00000381278    A4GALT ENSG00000128274         1
                     ...                 ...             ... |             ...             ...             ...       ...             ...       ...
      LRG_795t1    chr22 42720797 - 42720871               - |     LRG_795t1e1       LRG_795t1       LRG_795t1    A4GALT         LRG_795         1
      LRG_795t1    chr22 42695491 - 42695632               - |     LRG_795t1e2       LRG_795t1       LRG_795t1    A4GALT         LRG_795         2
      LRG_795t1    chr22 42692121 - 42693998               - |     LRG_795t1e3       LRG_795t1       LRG_795t1    A4GALT         LRG_795         3
------
seqinfo(1 sequences): chr22
 
Name:  
GenomicRanges with 26 ranges and 6 metadata columns
                seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
                   &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
ENST00000402174    chr22 28917389 - 28917482               + | ENSE00001552635 ENST00000402174 ENST00000402174     ZNRF3 ENSG00000183579         1
ENST00000402174    chr22 28987076 - 28987202               + | ENSE00003547707 ENST00000402174 ENST00000402174     ZNRF3 ENSG00000183579         2
ENST00000402174    chr22 29042495 - 29042570               + | ENSE00001329477 ENST00000402174 ENST00000402174     ZNRF3 ENSG00000183579         3
                     ...                 ...             ... |             ...             ...             ...       ...             ...       ...
ENST00000544604    chr22 29048389 - 29048492               + | ENSE00001325812 ENST00000544604 ENST00000544604     ZNRF3 ENSG00000183579         7
ENST00000544604    chr22 29049197 - 29050949               + | ENSE00001308098 ENST00000544604 ENST00000544604     ZNRF3 ENSG00000183579         8
ENST00000544604    chr22 29053579 - 29057489               + | ENSE00001427475 ENST00000544604 ENST00000544604     ZNRF3 ENSG00000183579         9
------
seqinfo(1 sequences): chr22
 
Name: ZNRF3 
GenomicRanges with 3 ranges and 6 metadata columns
                seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
                   &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
ENST00000325660    chr22 29031391 - 29031477               - | ENSE00001718965 ENST00000325660 ENST00000325660 ZNRF3-AS1 ENSG00000177993         1
ENST00000325660    chr22 29031084 - 29031157               - | ENSE00001643791 ENST00000325660 ENST00000325660 ZNRF3-AS1 ENSG00000177993         2
ENST00000325660    chr22 29024999 - 29027010               - | ENSE00001267818 ENST00000325660 ENST00000325660 ZNRF3-AS1 ENSG00000177993         3
------
seqinfo(1 sequences): chr22
 
Name: ZNRF3-AS1 
GenomicRanges with 2 ranges and 6 metadata columns
                seqnames              ranges          strand           exon_id         tx_name           tx_id gene_name         gene_id exon_rank
                   &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;            &lt;list&gt;          &lt;list&gt;          &lt;list&gt;    &lt;list&gt;          &lt;list&gt;    &lt;list&gt;
ENST00000412798    chr22 28992721 - 28992938               + | ENSE00001659048 ENST00000412798 ENST00000412798 ZNRF3-IT1 ENSG00000235786         1
ENST00000412798    chr22 29018164 - 29018621               + | ENSE00001595182 ENST00000412798 ENST00000412798 ZNRF3-IT1 ENSG00000235786         2
------
seqinfo(1 sequences): chr22
 
</code></pre>
</div>
</div>
</section>
<section id="finding-transcription-start-sites-tss" class="level3">
<h3 class="anchored" data-anchor-id="finding-transcription-start-sites-tss">2.2 Finding Transcription Start Sites (TSS)</h3>
<p>Transcription Start Sites (TSS) are the locations where transcription of a gene begins. Identifying TSS is crucial for understanding gene regulation, as many regulatory elements are located near the TSS.</p>
<p>Let’s use the <code>range()</code> method to get the full extent of each gene.</p>
<div id="cell-14" class="cell" data-scrolled="true" data-execution_count="26">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>ranges_by_gene <span class="op">=</span> by_gene.<span class="bu">range</span>()</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Gene ranges:"</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(ranges_by_gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Gene ranges:
GenomicRangesList with 932 ranges and 0 metadata columns
 
Name:  
GenomicRanges with 2 ranges and 0 metadata columns
    seqnames              ranges          strand
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
[0]    chr22 11066418 - 50674175               +
[1]    chr22 15282557 - 50755435               -
------
seqinfo(1 sequences): chr22
 
Name: 5_8S_rRNA 
GenomicRanges with 1 range and 0 metadata columns
    seqnames              ranges          strand
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
[0]    chr22 11249809 - 11249960               -
------
seqinfo(1 sequences): chr22
 
Name: A4GALT 
GenomicRanges with 1 range and 0 metadata columns
    seqnames              ranges          strand
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
[0]    chr22 42692121 - 42721299               -
------
seqinfo(1 sequences): chr22
 
Name:  
GenomicRanges with 1 range and 0 metadata columns
    seqnames              ranges          strand
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
[0]    chr22 28883572 - 29057489               +
------
seqinfo(1 sequences): chr22
 
Name: ZNRF3 
GenomicRanges with 1 range and 0 metadata columns
    seqnames              ranges          strand
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
[0]    chr22 29024999 - 29031477               -
------
seqinfo(1 sequences): chr22
 
Name: ZNRF3-AS1 
GenomicRanges with 1 range and 0 metadata columns
    seqnames              ranges          strand
       &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
[0]    chr22 28992721 - 29018621               +
------
seqinfo(1 sequences): chr22
 
</code></pre>
</div>
</div>
<p>We convert the list to a <code>GenomicRanges</code> object.</p>
<div id="cell-16" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>gr_by_gene <span class="op">=</span> ranges_by_gene.as_genomic_ranges()</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"as GenomicRanges:"</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr_by_gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>as GenomicRanges:
GenomicRanges with 936 ranges and 0 metadata columns
          seqnames              ranges          strand
             &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
             chr22 11066418 - 50674175               +
             chr22 15282557 - 50755435               -
5_8S_rRNA    chr22 11249809 - 11249960               -
               ...                 ...             ...
    ZNRF3    chr22 28883572 - 29057489               +
ZNRF3-AS1    chr22 29024999 - 29031477               -
ZNRF3-IT1    chr22 28992721 - 29018621               +
------
seqinfo(1 sequences): chr22</code></pre>
</div>
</div>
<p>Then we resize to a width of 1 base pair at the start of each range to pinpoint the TSS.</p>
<div id="cell-18" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>tss <span class="op">=</span> gr_by_gene.resize(width<span class="op">=</span><span class="dv">1</span>, fix<span class="op">=</span><span class="st">"start"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Transcript Start Sites:"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr_by_gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Transcript Start Sites:
GenomicRanges with 936 ranges and 0 metadata columns
          seqnames              ranges          strand
             &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
             chr22 11066418 - 50674175               +
             chr22 15282557 - 50755435               -
5_8S_rRNA    chr22 11249809 - 11249960               -
               ...                 ...             ...
    ZNRF3    chr22 28883572 - 29057489               +
ZNRF3-AS1    chr22 29024999 - 29031477               -
ZNRF3-IT1    chr22 28992721 - 29018621               +
------
seqinfo(1 sequences): chr22</code></pre>
</div>
</div>
</section>
<section id="defining-promoter-regions" class="level3">
<h3 class="anchored" data-anchor-id="defining-promoter-regions">2.3 Defining Promoter Regions</h3>
<p>Here, we’re defining promoters as the region 2000 base pairs upstream to 200 base pairs downstream of each TSS. This definition can vary depending on the specific analysis, but this range often captures important regulatory elements.</p>
<div id="cell-20" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>promoters <span class="op">=</span> tss.promoters(upstream<span class="op">=</span><span class="dv">2000</span>, downstream<span class="op">=</span><span class="dv">200</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Promoter Regions:"</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(promoters)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Promoter Regions:
GenomicRanges with 936 ranges and 0 metadata columns
          seqnames              ranges          strand
             &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
             chr22 11064418 - 11066618               +
             chr22 50755235 - 50757435               -
5_8S_rRNA    chr22 11249760 - 11251960               -
               ...                 ...             ...
    ZNRF3    chr22 28881572 - 28883772               +
ZNRF3-AS1    chr22 29031277 - 29033477               -
ZNRF3-IT1    chr22 28990721 - 28992921               +
------
seqinfo(1 sequences): chr22</code></pre>
</div>
</div>
<pre class="{note}"><code>Please be aware that because gene symbols may not be unique, this `GenomicRanges` object might contain duplicates. You might want to resolve duplicate symbols by making the symbols unique. We will leave this as an exercise for the reader.</code></pre>
</section>
</section>
<section id="overlap-analysis" class="level2">
<h2 class="anchored" data-anchor-id="overlap-analysis">3. Overlap Analysis</h2>
<p>A common task in genomic analysis is finding overlaps between different genomic features. This helps us understand the relationships between various elements in the genome and can provide insights into gene regulation and function.</p>
<section id="reading-chip-seq-peaks" class="level3">
<h3 class="anchored" data-anchor-id="reading-chip-seq-peaks">3.1 Reading ChIP-seq Peaks</h3>
<p>ChIP-seq (Chromatin Immunoprecipitation followed by sequencing) is a method used to identify binding sites of DNA-associated proteins. The peaks represent regions where a protein of interest is likely bound to the DNA. We’re focusing on chromosome 22 for this example to keep the dataset manageable.</p>
<p>For the purpose of this tutorial, let’s download a bed file containing peaks from a ChIP-seq experiment on “Human B cells” to identify “EZH2” binding sites (from ENCODE) and cataloged in <a href="https://bedbase.org/bed/be4054acf6e3feeb4dc490e6430e358e">bedbase.org</a>.</p>
<div id="cell-22" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> geniml.bbclient <span class="im">import</span> BBClient</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>bbclient <span class="op">=</span> BBClient(cache_folder<span class="op">=</span><span class="st">"cache"</span>, bedbase_api<span class="op">=</span><span class="st">"https://api.bedbase.org"</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>bedfile_id <span class="op">=</span> <span class="st">"be4054acf6e3feeb4dc490e6430e358e"</span> </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>bedfile <span class="op">=</span> bbclient.load_bed(bedfile_id)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>peaks <span class="op">=</span> bedfile.to_granges()</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>filter_chr22 <span class="op">=</span> [x <span class="op">==</span> <span class="st">"chr22"</span> <span class="cf">for</span> x <span class="kw">in</span> peaks.get_seqnames()]</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>peaks_chr22 <span class="op">=</span> peaks[filter_chr22]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(peaks_chr22)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GenomicRanges with 1441 ranges and 0 metadata columns
       seqnames              ranges          strand
          &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
   [0]    chr22 19766788 - 19767078               *
   [1]    chr22 17369888 - 17370178               *
   [2]    chr22 19756445 - 19756735               *
            ...                 ...             ...
[1438]    chr22 27212058 - 27212348               *
[1439]    chr22 49201359 - 49201649               *
[1440]    chr22 49663362 - 49663652               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX</code></pre>
</div>
</div>
</section>
<section id="finding-overlaps-with-tss" class="level3">
<h3 class="anchored" data-anchor-id="finding-overlaps-with-tss">3.2 Finding Overlaps with TSS</h3>
<p>Here, we’re identifying ChIP-seq peaks that overlap with TSS. This analysis can help us understand if the protein of interest tends to bind near the start of genes, which could suggest a role in transcription initiation.</p>
<div id="cell-24" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>overlaps <span class="op">=</span> peaks_chr22.find_overlaps(tss)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Peak indices that overlap with first 10 TSS:"</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(overlaps[:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Peak indices that overlap with first 10 TSS:
[[], [], [], [55], [217], [], [], [], [], []]</code></pre>
</div>
</div>
<pre class="{note}"><code>`find_overlaps` returns a `list` with the same length as TSS, indicating which indices from peaks overlap with each of the TSS. Ideally, we would want to return a `Hits` object similar to the Bioconductor implementation.

**TODO: Future plans to convert this into a `Hits` object.**</code></pre>
<p>Let’s identify the peaks that overlap with TSS.</p>
<div id="cell-26" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> itertools</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>all_indices <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(itertools.chain.from_iterable(overlaps)))</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>peaks_by_tss <span class="op">=</span> peaks_chr22[all_indices]</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(peaks_by_tss)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GenomicRanges with 35 ranges and 0 metadata columns
     seqnames              ranges          strand
        &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
 [0]    chr22 19756445 - 19756735               *
 [1]    chr22 38467935 - 38468225               *
 [2]    chr22 24952664 - 24952954               *
          ...                 ...             ...
[32]    chr22 21032552 - 21032842               *
[33]    chr22 50270553 - 50270843               *
[34]    chr22 19131257 - 19131547               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX</code></pre>
</div>
</div>
<p>Instead, one can subset peaks that overlap with TSS using the <code>subset_by_overlaps</code> method:</p>
<div id="cell-28" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>peaks_by_tss2 <span class="op">=</span> peaks_chr22.subset_by_overlaps(tss)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(peaks_by_tss2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GenomicRanges with 35 ranges and 0 metadata columns
     seqnames              ranges          strand
        &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
 [0]    chr22 19756445 - 19756735               *
 [1]    chr22 38467935 - 38468225               *
 [2]    chr22 24952664 - 24952954               *
          ...                 ...             ...
[32]    chr22 21032552 - 21032842               *
[33]    chr22 50270553 - 50270843               *
[34]    chr22 19131257 - 19131547               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX</code></pre>
</div>
</div>
<p>Additionally, in some cases, we may want to ignore strand information (<code>ignore_strand=True</code>) when finding overlaps.</p>
<div id="cell-30" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>peaks_by_tss_ignoring_strand <span class="op">=</span> peaks_chr22.subset_by_overlaps(tss, ignore_strand<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(peaks_by_tss_ignoring_strand)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GenomicRanges with 35 ranges and 0 metadata columns
     seqnames              ranges          strand
        &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
 [0]    chr22 19756445 - 19756735               *
 [1]    chr22 38467935 - 38468225               *
 [2]    chr22 24952664 - 24952954               *
          ...                 ...             ...
[32]    chr22 21032552 - 21032842               *
[33]    chr22 50270553 - 50270843               *
[34]    chr22 19131257 - 19131547               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX</code></pre>
</div>
</div>
<pre class="{note}"><code>This yields the same results for this particular scenario, but may not if the 'peaks' contain strand information.</code></pre>
</section>
<section id="finding-overlaps-with-promoters" class="level3">
<h3 class="anchored" data-anchor-id="finding-overlaps-with-promoters">3.3 Finding Overlaps with Promoters</h3>
<p>This operation finds ChIP-seq peaks that overlap with our defined promoter regions. If a significant number of peaks fall within promoters, it might suggest that the protein plays a role in gene regulation through promoter binding. This kind of analysis is often used to characterize the binding patterns of transcription factors or other regulatory proteins.</p>
<div id="cell-32" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>peaks_by_promoters <span class="op">=</span> peaks_chr22.subset_by_overlaps(promoters)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Peaks Overlapping with Promoters:"</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(peaks_by_promoters)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Peaks Overlapping with Promoters:
GenomicRanges with 190 ranges and 0 metadata columns
      seqnames              ranges          strand
         &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
  [0]    chr22 19756445 - 19756735               *
  [1]    chr22 37427967 - 37428257               *
  [2]    chr22 22521942 - 22522232               *
           ...                 ...             ...
[187]    chr22 39993439 - 39993729               *
[188]    chr22 22338004 - 22338294               *
[189]    chr22 19131257 - 19131547               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX</code></pre>
</div>
</div>
</section>
<section id="finding-overlaps-with-exons" class="level3">
<h3 class="anchored" data-anchor-id="finding-overlaps-with-exons">3.4 Finding Overlaps with Exons</h3>
<p>Another analysis is to look at overlaps with all exons. This can help identify potential roles of the ChIP-seq peaks in splicing. Let’s modify our analysis to look at all exons:</p>
<div id="cell-34" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine all exons into a single GenomicRanges object</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>all_exons <span class="op">=</span> by_gene.as_granges()</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Find peaks overlapping with any exon</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>peaks_by_exons <span class="op">=</span> peaks_chr22.subset_by_overlaps(all_exons)</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Peaks overlapping with exons:"</span>)</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(peaks_by_exons)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the percentage of peaks that overlap with exons</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>percent_overlapping <span class="op">=</span> (<span class="bu">len</span>(peaks_by_exons) <span class="op">/</span> <span class="bu">len</span>(peaks_chr22)) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Percentage of peaks overlapping with exons: </span><span class="sc">{</span>percent_overlapping<span class="sc">:.2f}</span><span class="ss">%"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Peaks overlapping with exons:
GenomicRanges with 279 ranges and 0 metadata columns
      seqnames              ranges          strand
         &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
  [0]    chr22 19766788 - 19767078               *
  [1]    chr22 17369888 - 17370178               *
  [2]    chr22 29307104 - 29307394               *
           ...                 ...             ...
[276]    chr22 16969920 - 16970210               *
[277]    chr22 35552420 - 35552710               *
[278]    chr22 37931897 - 37932187               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX
Percentage of peaks overlapping with exons: 19.36%</code></pre>
</div>
</div>
<p>This analysis can provide insights into whether the protein of interest (captured by the ChIP-seq: “EZH2”) tends to bind within gene bodies, potentially influencing gene expression, splicing, or other co-transcriptional processes.</p>
</section>
</section>
<section id="advanced-operations" class="level2">
<h2 class="anchored" data-anchor-id="advanced-operations">4. Advanced Operations</h2>
<p>Let’s explore some more complex operations that are often used in genomic analyses.</p>
<section id="comparing-exonic-vs.-intronic-binding" class="level3">
<h3 class="anchored" data-anchor-id="comparing-exonic-vs.-intronic-binding">4.1 Comparing Exonic vs.&nbsp;Intronic Binding</h3>
<p>Let’s first identify intron regions. We will use the <code>by_gene</code> object we created that contains a <code>GenomicRangesList</code> split by gene.</p>
<div id="cell-36" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create intronic regions (regions within genes but not in exons)</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>gene_ranges <span class="op">=</span> by_gene.<span class="bu">range</span>().as_genomic_ranges()  <span class="co"># Get the full extent of each gene</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>introns <span class="op">=</span> gene_ranges.subtract(all_exons).as_granges()</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Intron regions:"</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(introns)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Intron regions:
GenomicRanges with 1572 ranges and 0 metadata columns
          seqnames              ranges          strand
             &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
             chr22 15282557 - 15550904               -
             chr22 15552888 - 15553211               -
             chr22 15553587 - 15553691               -
               ...                 ...             ...
    ZNRF3    chr22 28883572 - 29057489               +
ZNRF3-AS1    chr22 29024999 - 29031477               -
ZNRF3-IT1    chr22 28992721 - 29018621               +
------
seqinfo(1 sequences): chr22</code></pre>
</div>
</div>
<p>To gain further insight, we can compare the proportion of peaks overlapping with exons to those overlapping with introns:</p>
<div id="cell-38" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find peaks overlapping with introns</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>peaks_by_introns <span class="op">=</span> peaks_chr22.subset_by_overlaps(introns)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Peaks overlapping with introns:"</span>)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(peaks_by_introns)</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate percentages</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>percent_exonic <span class="op">=</span> (<span class="bu">len</span>(peaks_by_exons) <span class="op">/</span> <span class="bu">len</span>(peaks_chr22)) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>percent_intronic <span class="op">=</span> (<span class="bu">len</span>(peaks_by_introns) <span class="op">/</span> <span class="bu">len</span>(peaks_chr22)) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Percentage of peaks overlapping with exons: </span><span class="sc">{</span>percent_exonic<span class="sc">:.2f}</span><span class="ss">%"</span>)</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Percentage of peaks overlapping with introns: </span><span class="sc">{</span>percent_intronic<span class="sc">:.2f}</span><span class="ss">%"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Peaks overlapping with introns:
GenomicRanges with 1438 ranges and 0 metadata columns
       seqnames              ranges          strand
          &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
   [0]    chr22 19766788 - 19767078               *
   [1]    chr22 17369888 - 17370178               *
   [2]    chr22 19756445 - 19756735               *
            ...                 ...             ...
[1435]    chr22 27212058 - 27212348               *
[1436]    chr22 49201359 - 49201649               *
[1437]    chr22 49663362 - 49663652               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX
Percentage of peaks overlapping with exons: 19.36%
Percentage of peaks overlapping with introns: 99.79%</code></pre>
</div>
</div>
<pre class="{note}"><code>These percentages add up to over 100% because some peaks overlap both introns and exons, depending on how wide the peaks are. Ideally, you may want to filter the peaks based on preference as you annotate them with TSS, promoters, etc.</code></pre>
<p>This comparison can help determine if the protein of interest shows a preference for binding in exonic or intronic regions, which could suggest different functional roles (e.g., splicing regulation for exonic binding vs.&nbsp;potential enhancer activity for intronic binding).</p>
</section>
<section id="finding-overlaps-with-the-first-exon" class="level3">
<h3 class="anchored" data-anchor-id="finding-overlaps-with-the-first-exon">4.2 Finding Overlaps with the first exon</h3>
<pre class="{note}"><code>- This analysis is performed by transcript.
- The rationale for this analysis may vary, but we are mostly showcasing complex genomic operations that are possible with the package.</code></pre>
<p>Let’s first put together a <code>GenomicRanges</code> object containing the first exon for each transcript.</p>
<div id="cell-40" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>all_first <span class="op">=</span> []</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> txid, grl <span class="kw">in</span> by_tx:</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    strand <span class="op">=</span> grl.get_strand(as_type <span class="op">=</span> <span class="st">"list"</span>)[<span class="dv">0</span>]</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> strand <span class="op">==</span> <span class="st">"-"</span>:</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>        all_first.append(grl.sort()[<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>        all_first.append(grl.sort()[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we combine all the individual genomic elements. The <a href="https://github.com/BiocPy/BiocUtils">biocutils</a> package provides utilities for convenient aspects of R that aren’t provided by base Python and generics. One of these generics is the <code>'combine'</code> operation that merges or concatenates various Bioconductor classes.</p>
<div id="cell-42" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> biocutils <span class="im">import</span> combine_sequences</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>first_exons <span class="op">=</span> combine_sequences(<span class="op">*</span>all_first)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now subset peaks that overlap with the first exon</p>
<div id="cell-44" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>peaks_with_first_exons <span class="op">=</span> peaks_chr22.subset_by_overlaps(first_exons)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(peaks_with_first_exons)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GenomicRanges with 153 ranges and 0 metadata columns
      seqnames              ranges          strand
         &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
  [0]    chr22 17369888 - 17370178               *
  [1]    chr22 19756445 - 19756735               *
  [2]    chr22 45975507 - 45975797               *
           ...                 ...             ...
[150]    chr22 49500975 - 49501265               *
[151]    chr22 19131257 - 19131547               *
[152]    chr22 29307104 - 29307394               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX</code></pre>
</div>
</div>
</section>
<section id="resizing-and-shifting-peaks" class="level3">
<h3 class="anchored" data-anchor-id="resizing-and-shifting-peaks">4.3 Resizing and Shifting Peaks</h3>
<div id="cell-46" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>narrow_peaks <span class="op">=</span> peaks_chr22.narrow(start<span class="op">=</span><span class="dv">10</span>, width<span class="op">=</span><span class="dv">100</span>)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>shifted_peaks <span class="op">=</span> narrow_peaks.shift(<span class="dv">10</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Narrowed and Shifted Peaks:"</span>)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(shifted_peaks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Narrowed and Shifted Peaks:
GenomicRanges with 1441 ranges and 0 metadata columns
       seqnames              ranges          strand
          &lt;str&gt;           &lt;IRanges&gt; &lt;ndarray[int8]&gt;
   [0]    chr22 19766807 - 19766907               *
   [1]    chr22 17369907 - 17370007               *
   [2]    chr22 19756464 - 19756564               *
            ...                 ...             ...
[1438]    chr22 27212077 - 27212177               *
[1439]    chr22 49201378 - 49201478               *
[1440]    chr22 49663381 - 49663481               *
------
seqinfo(46 sequences): chr1 chr10 chr11 ... chrUn_KI270750v1 chrUn_KI270752v1 chrX</code></pre>
</div>
</div>
<p>Resizing and shifting genomic ranges can be useful in various contexts. For example:</p>
<ul>
<li>Narrowing peaks might help focus on the center of ChIP-seq binding sites.</li>
<li>Shifting ranges can be used to look at regions adjacent to your features of interest. e.g.&nbsp;defining the predicted CRISPR cleavage site based on the position of the CRISPR gRNA sequence.</li>
</ul>
<p>These operations demonstrate the flexibility of genomic range manipulations, which can be useful for fine-tuning analyses or testing hypotheses about the spatial relationships between genomic features.</p>
</section>
</section>
<section id="exercises" class="level2">
<h2 class="anchored" data-anchor-id="exercises">5. Exercises</h2>
<ol type="1">
<li>Calculate the average width of the ChIP-seq peaks on chromosome 22.</li>
<li>Determine how many peaks overlap with CpG islands.</li>
<li>Compute the percentage of promoter regions that have at least one overlapping ChIP-seq peak.</li>
</ol>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>In this tutorial, we’ve explored how to use BiocPy’s genomic ranges functionality to perform various genomic analyses. These tools and techniques provide a powerful way to work with genomic interval data in Python, mirroring the capabilities from Bioconductor. They form the foundation for many more complex genomic analyses and can be applied to a wide range of biological questions.</p>
<pre class="{note}"><code>Refer to the [BiocPy documentation](https://biocpy.github.io/) for more detailed information on these packages and their functionalities.</code></pre>


</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
 (c) Jayaram Kancherla &amp; Aaron Lun.
  </li>  
</ul>
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/BiocPy/BiocWorkshop2024/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>