# Welcome

Welcome to our workshop on exploring the data structures and packages 
available in [BiocPy](https://github.com/biocpy), a project that aims 
to facilitate Bioconductor workflows in Python. 

In this workshop, we will focus on interoperability between R and Python, covering two main topics:

- Reading an RDS file containing `GenomicRanges` and performing downstream range-based analyses.
- Annotating single-cell RNA-seq data using the [scrnaseq](https://github.com/biocpy/scrnaseq) package.

Attendees will learn how to represent and manipulate their datasets in Python 
in the same manner as in R/Bioconductor. 
All packages in BiocPy are published to PyPI, and the code is open-source 
on [GitHub](https://github.com/BiocPy).


### Core contributors

- [Jayaram Kancherla](https://github.com/jkanche)
- [Aaron Lun](https://github.com/LTLA)

We are always looking for more contributions from the community to improve our packages! Please check out the issues or discussion in our GitHub organization if you are interested.

----

## Other resources

- A more detailed tutorial on BiocPy representations is available [here](https://biocpy.github.io/tutorial/).
- Check out all the package in the BiocPy ecosystem on [GitHub](https://github.com/biocpy).

## Developer notes

This is a reproducible Quarto book with reusable snippets. To learn more about Quarto books visit <https://quarto.org/docs/books>. Check out [Session Info](./chapters/sessioninfo.qmd) for more information.